lib/rust/profiling/src/lib.rs

```diff
@@ -0,0 +1,438 @@
+use ::macros::*;
+
+use enso_prelude::fmt::Formatter;
+use enso_web::performance;
+use inflector::Inflector;
+use ordered_float::OrderedFloat;
+use serde::Deserialize;
+use serde::Serialize;
+use serde_plain::from_str;
+use wasm_bindgen::JsValue;
+use web_sys::PerformanceEntry;
+
+
+
+// ================
+// === Metadata ===
+// ================
+
+/// Source code location given as file path and line number.
+#[derive(Serialize, Deserialize, Clone, Debug)]
+#[allow(missing_docs)]
+pub struct SourceLocation {
+    pub file: String,
+    pub line: u32,
+}
+
+/// Measurement metadata. This struct holds information about a measurement and provides
+/// functionality for conversion form/to JS for use in the performance API.
+#[derive(Serialize, Deserialize, Clone, Debug)]
+pub struct Metadata {
+    /// Source code location of the profiling interval.
+    pub source:          SourceLocation,
+    /// Profiling level of the measurement.
+    pub profiling_level: ProfilingLevel,
+    /// Label of the measurement..
+    pub label:           String,
+}
+
+impl From<Metadata> for JsValue {
+    fn from(metadata: Metadata) -> JsValue {
+        JsValue::from_serde(&metadata).expect("Failed to serialise Metadata struct to JSON.")
+    }
+}
+
+impl TryFrom<JsValue> for Metadata {
+    type Error = serde_json::Error;
+
+    fn try_from(value: JsValue) -> Result<Self, Self::Error> {
+        value.into_serde()
+    }
+}
+
+
+
+// =================================
+// === Custom JS Performance API ===
+// =================================
+
+mod js {
+    use super::*;
+    use js_sys::JsString;
+    use wasm_bindgen::JsValue;
+
+    #[wasm_bindgen(inline_js = "
+export function mark_with_metadata(markName, markOptions) {
+   performance.mark(markName, markOptions)
+}
+
+export function measure_with_start_mark_and_end_mark_and_metadata(
+    measureName,
+    startMark,
+    endMark,
+    measureOptions
+) {
+    const options = {}
+    options.start = startMark
+    options.end = endMark
+    options.detail = measureOptions
+    performance.measure(measureName, options)
+}
+
+")]
+    extern "C" {
+        #[allow(unsafe_code)]
+        pub fn mark_with_metadata(mark_name: JsString, mark_options: JsValue);
+        #[allow(unsafe_code)]
+        pub fn measure_with_start_mark_and_end_mark_and_metadata(
+            measure_name: JsString,
+            start_mark: JsString,
+            end_mark: JsString,
+            measure_options: JsValue,
+        );
+    }
+}
+
+
+
+// =================
+// === Profilers ===
+// =================
+
+#[allow(missing_docs)]
+type ProfilingLevel = String;
+
+define_profiler! {
+    $, "section", Section, section, start_section, end_section, measure_section;
+    $, "task", Task, task, start_task, end_task, measure_task;
+    $, "detail", Detail, detail, start_detail, end_detail, measure_detail;
+    $, "debug", Debug, debug, start_debug, end_debug, measure_debug;
+}
+
+define_hierarchy!(section, task, detail, debug);
```

Comment:
you could do it from one macro to repeat things less. Also, remember that you have the `stringify!` macro in Rust which can convert any macro token to string literal for you :)